### PR TITLE
Add automatic retry interceptor for HTTP 429

### PR DESCRIPTION
**Pull request by milismsft**
Add automatic retry interceptor for HTTP 429: Azure Resource Manager throttling when read/write per hour request limit reached.

### CODE REVIEW

File: azure-mgmt-resources/src/main/java/com/microsoft/azure/management/resources/fluentcore/utils/ResourceManagerThrottlingInterceptor.java [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+            synchronized (REENTRANT_LOCK) {
+                if (REENTRANT_LOCK.isLocked()) {
+                    REENTRANT_LOCK.wait();
+                }
+            }
+        } catch (InterruptedException e) {
+            throw new IOException(e);
+        }
+        Response response = chain.proceed(chain.request());
+        if (response.code() != 429) {
```
Comment (martinsawicki) [inline azure-mgmt-resources/src/main/java/com/microsoft/azure/management/resources/fluentcore/utils/ResourceManagerThrottlingInterceptor.java]: is `response` guaranteed to be never null here?  (same for `chain`)

File: azure-mgmt-resources/src/main/java/com/microsoft/azure/management/resources/fluentcore/utils/ResourceManagerThrottlingInterceptor.java [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+            String retryAfterHeader = response.header("Retry-After");
+            int retryAfter = 0;
+            if (retryAfterHeader != null) {
+                retryAfter = Integer.parseInt(retryAfterHeader);
+            }
+            if (retryAfter <= 0) {
+                Pattern pattern = Pattern.compile("try again after '([0-9]*)' minutes", Pattern.CASE_INSENSITIVE);
+                Matcher matcher = pattern.matcher(content(response.body()));
+                if (matcher.find()) {
+                    retryAfter = (int) TimeUnit.MINUTES.toSeconds(Integer.parseInt(matcher.group(1)));
```
Comment (martinsawicki) [inline azure-mgmt-resources/src/main/java/com/microsoft/azure/management/resources/fluentcore/utils/ResourceManagerThrottlingInterceptor.java]: `matcher.group(1)` could return null theoretically

File: azure-mgmt-resources/src/main/java/com/microsoft/azure/management/resources/fluentcore/utils/ResourceManagerThrottlingInterceptor.java [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+
+        try {
+            String retryAfterHeader = response.header("Retry-After");
+            int retryAfter = 0;
+            if (retryAfterHeader != null) {
+                retryAfter = Integer.parseInt(retryAfterHeader);
+            }
+            if (retryAfter <= 0) {
+                Pattern pattern = Pattern.compile("try again after '([0-9]*)' minutes", Pattern.CASE_INSENSITIVE);
+                Matcher matcher = pattern.matcher(content(response.body()));
```
Comment (martinsawicki) [inline azure-mgmt-resources/src/main/java/com/microsoft/azure/management/resources/fluentcore/utils/ResourceManagerThrottlingInterceptor.java]: shouldn't we protect against `response.body()` returning null?

File: azure-mgmt-resources/src/main/java/com/microsoft/azure/management/resources/fluentcore/utils/ResourceManagerThrottlingInterceptor.java [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+                REENTRANT_LOCK.notifyAll();
+            }
+        }
+    }
+
+    private String content(ResponseBody responseBody) throws IOException {
+        if (responseBody == null) {
+            return null;
+        }
+        BufferedSource source = responseBody.source();
```
Comment (martinsawicki) [inline azure-mgmt-resources/src/main/java/com/microsoft/azure/management/resources/fluentcore/utils/ResourceManagerThrottlingInterceptor.java]: is this guaranteed to not be null?

File: azure-mgmt-resources/src/main/java/com/microsoft/azure/management/resources/fluentcore/utils/ResourceManagerThrottlingInterceptor.java [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+        source.request(Long.MAX_VALUE); // Buffer the entire body.
+        Buffer buffer = source.buffer();
```
Comment (martinsawicki) [inline azure-mgmt-resources/src/main/java/com/microsoft/azure/management/resources/fluentcore/utils/ResourceManagerThrottlingInterceptor.java]: null check?

File: azure-mgmt-resources/src/main/java/com/microsoft/azure/management/resources/fluentcore/utils/ResourceManagerThrottlingInterceptor.java (hunk 1 above)
Comment (milismsft) [inline azure-mgmt-resources/src/main/java/com/microsoft/azure/management/resources/fluentcore/utils/ResourceManagerThrottlingInterceptor.java]: yes; chain.proceed will through an NullPointerException if it was.

File: azure-mgmt-resources/src/main/java/com/microsoft/azure/management/resources/fluentcore/utils/ResourceManagerThrottlingInterceptor.java (hunk 3 above)
Comment (milismsft) [inline azure-mgmt-resources/src/main/java/com/microsoft/azure/management/resources/fluentcore/utils/ResourceManagerThrottlingInterceptor.java]: we will always have a body in the HTTP response even if that might be empty (but not null).

File: azure-mgmt-resources/src/main/java/com/microsoft/azure/management/resources/fluentcore/utils/ResourceManagerThrottlingInterceptor.java (hunk 2 above)
Comment (milismsft) [inline azure-mgmt-resources/src/main/java/com/microsoft/azure/management/resources/fluentcore/utils/ResourceManagerThrottlingInterceptor.java]: not the case here; matcher grouping starts at index 1 per Java Matcher docs.

File: azure-mgmt-resources/src/main/java/com/microsoft/azure/management/resources/fluentcore/utils/ResourceManagerThrottlingInterceptor.java (hunk 4 above)
Comment (milismsft) [inline azure-mgmt-resources/src/main/java/com/microsoft/azure/management/resources/fluentcore/utils/ResourceManagerThrottlingInterceptor.java]: yes, unless Azure returns a malformed HTTP response...

File: azure-mgmt-resources/src/main/java/com/microsoft/azure/management/resources/fluentcore/utils/ResourceManagerThrottlingInterceptor.java (hunk 5 above)
Comment (milismsft) [inline azure-mgmt-resources/src/main/java/com/microsoft/azure/management/resources/fluentcore/utils/ResourceManagerThrottlingInterceptor.java]: same as above.

**Comment by milismsft** [comment]
I'm ok to bullet proof with null checks on the relevant response properties but I don't think it is necessary because we will always have a valid RealSesponceBody object behind it that is not null (it's source might be an empty buffer at most).

**Comment by milismsft** [comment]
BTW for the sake of making sure we don't blow up, I removed the body from one of the 429 responses in the JSON recording file and the test still passes without issues (I had to disable one other interceptor though that expects a valid body).

**Pull request merged**